### DISCUSSION TITLE
posts/programming/nest

### DISCUSSION THREAD

**Discussion by giscus** [movie42/hyunsu#11] [General]
# posts/programming/nest

이 글은 내가 docker 없이 배포하는 방법을 기억하기 위해 쓴 글이라 누군가가 읽었을 때 아무것도 얻어갈 수 없을 것이다.

# NestJS

express, fastify로 이어서 NestJS까지 어플리케이션을 만들어보았다. 하지만 대부분 프론트 어플리케이션을 동작시키기 위한 용도로 배운거라
깊게 무언가 익혔다고 보긴 어려운 것 같다.

## NestJS 도움되는 강의

[따라하며 배우는 NestJS](https://www.inflearn.com/course/%EB%94%B0%EB%9D%BC%ED%95%98%EB%8A%94-%EB%84%A4%EC%8A%A4%ED%8A%B8-%EC%A0%9C%EC%9D%B4%EC%97%90%EC%8A%A4/dashboard)

# 배포

## 클라우드

다 만들었다면 배포를 하자. 배포는 클라우드 컴퓨팅을 사용한다. 서버는 어렵다고 생각할 수도 있지만 그냥 컴퓨터다. 남의 컴퓨터를 빌려쓰는 거라고 생각하자.
빌려쓰려면 돈을 내야한다. 하지만 무료로 최소 사양 컴퓨터를 영구적으로 빌려 쓸 수 있다. 언젠가 사라질 수 있는 서비스이지만 무료인데 안쓸 이유가 있나?

- [오라클 클라우드](https://www.oracle.com/)

티어 한개는 무제한 무료다.

클라우드를 만드는 방법은 AWS나 Oracle이나 인터페이스의 차이만 있을 뿐 거의 같다. 아래 동영상을 참고하자. 나는 ubuntu가 편해서 그냥 ubuntu로 했다. 초보라면 동영상을 그대로 따라하고 모르는 단어나 개념을 GPT에게 물어보자.

- [Tip&Tech - Linux, MySQL 오라클 클라우드 무료 서버 사용하기, @시코-시니어코딩](https://www.youtube.com/watch?v=wLgHEn2vcPo)

## 내가 만든 프로젝트 가져오기

git을 설치한 다음 내 프로젝트를 클론해오자.

```shell
git clone project name
```

## node 설치

이제 node가 필요하다. 그런데 apt로 node를 설치하면 v10.\*이 설치된다. 자 다시 지우자.

그리고 [NodeSource Node.js Binary Distributions
](https://github.com/nodesource/distributions/blob/master/README.md)를 들어가서 가이드대로 다시 설치하자.

하지만 이 방법이 유일한건 아니다. 아래 방법도 있으니까 참고해보자.

[Ubuntu에 Node.js를 설치하고 npm을 최신 버전으로 업데이트하는 방법](https://www.freecodecamp.org/korean/news/how-to-install-node-js-on-ubuntu-and-update-npm-to-the-latest-version/)

그리고 yarn을 설치하고 패키지를 설치한다. 패키지 설치가 끝나면 build를 한다. build가 끝나고 build한 파일 폴더로 가서 build된 어플을 실행해보자.

```shell
node main.js
```

에러가 엄청 날텐데 두 가지를 해결해야한다.

1. 환경변수
2. db

## db 설치하기

나는 postgresql을 설치했다. 설치를 다 했다면 postgresql을 실행하자.

```shell
systemctl start postgresql
```

```shell
systemctl status postgresql
```

그 다음 postgres에 접속한다.

```shell
sudo -i -u postgres
```

그럼 postgres에 접속하게 된다. 아래 psql을 입력하자.

```shell
postgres@name:~$ psql
```

db 만들기, user 만들기 권한주기 등은 gpt에게 물어보면 잘 알려준다.

db 세팅이 끝나면 [dbeaver](https://dbeaver.io/download/)를 설치한 다음에 연결해주자.
dbeaver 연결할 때, ssh로 연결해주고 main 설정까지 해주어야한다. (로컬이랑 연결 방법이 좀 다름)

## 환경변수 세팅하기

```shell
vi .bashrc
```

password를 1234로 지정하는 정신나간 짓은 안할꺼라 생각한다. 여긴 그냥 예제라 이렇게 보여주었다.

```
export DB_HOST=localhost
export DB_NAME=writer
export DB_USER=admin
export DB_PASSWORD=1234
```

저장하고 종료한다.

# 다시 build

build 폴더를 지우고 다시 build하자.

```shell
rm -rf dist
yarn build
```

# 앞으로 해결 과제

1. 이전에 mongodb로 할때는 migration 같은게 필요 없었는데 sql은 다른 것 같다. 일단 migration 명령어를 넣어놓지 않아서
   dev로 실행을 한번 한다음에 그 다음에 실행하니까 된다.
2. 로컬에서는 signup하고 login한 다음 return되는 jwt 토큰을 authrization에 붙여넣기 해서 boards를 실행하면 게시물이 만들어지는데 이상하게 프러덕션에서는 401에러가 뜬다. 왜 권한이 없다고 그러는지 이해가 안된다.
   이것때문에 3시간 하다가 결국 못했다. 같은 도메인으로 인식하지 못하는건가? 게시물 생성 결국 실패했다.
3. 오라클은 FQDN이 내부 FQDN만 주어져서 그런지 외부에서 이 DNS로 접속이 안된다. IP를 외부에 노출하고 싶지 않다. DNS를 구매해야하나? 어차피 1년이면 싼거 구매 가능한데 찾아봐야겠다.
4. 만약 서버에 새로운 것이 추가되었다고 생각해보자. CI/CD 설정 안되어있으면 진짜 끔찍하다. 배포 파이프라인 구축도 알아보자.

# 아직 마무리는 아니다.

스터디에서 프로젝트를 만들어보자고 시작하게 된건데 와... 진짜 서버 ME(mongodb, expressjs) 스택으로 할때랑 뭐가 많이 다르다. 간단하지가 않다.
PN(postgresql, nestjs)스택은 DB 마이그레이션이라는 개념도 익숙치 않고 Service, Repository, Controller, Module 로 코드가 있어야할 위치에 있도록 정해놓아서 자유도는 떨어져도
express로 할때보다 덜 혼란스러운것도 사실이다.(물론 express도 그냥 이런 규칙을 정하고 폴더구조를 잡고 코드를 작성하면 된다.)
하지만 이정도로는 NestJS의 장점을 잘 모르겠다. 물론 데코레이션을 사용해서 validation을 한다던가 하는건 정말 편하다. DTO를 만든다던가 하는건 정말 편하다.
사실 요즘 bun이 나오면서 hono, elisia 등 여러 종류의 서버 프레임워크가 쏟아져 나오는데
그런데 ExpressJS가 정말 대단한게 왠만한 JS 진영의 서버 프레임워크는 거의 다 express에 영감을 받았다고 말한다.(진짜 GOAT...)
아무튼 그 상황에서 NestJS를 선택한건 사실 좀 쉽게 가자는게 컸는데(서버 빨리 만들어놓고 React 18로 프로젝트 진행해보면서 React 18을 익히자는 의도가 컸다.) 쉽지 않다.

이 아티클은 새로운 것을 익힐 때마다 계속 업데이트 할 예정이다.

고럼 이만.

https://www.hyunsu.info/posts/programming/nest

**Comment by movie42** [comment]
# Unauthrization 찾은 이슈들
참조해보기(내 경우에는 github 이슈랑 비슷한 듯 하다. 하지만 local에서는 잘 동작하는데 prd만 가면 동작을 안하는게 이상하다. 아마 설치할때 버전 문제일 수 도 있을 것 같다.)
[stackoverflow](https://stackoverflow.com/questions/62799708/nest-js-auth-guard-jwt-authentication-constantly-returns-401-unauthorized)
[github](https://github.com/nestjs/docs.nestjs.com/issues/875#issuecomment-619472086)
